### PR TITLE
QT: bump fee returns PSBT on clipboard for watchonly-only wallets

### PR DESCRIPTION
**Pull request by instagibbs**
Very small set of changes to support PSBT-based fee bumping on watchonly wallets in QT.

quasi-companion to https://github.com/bitcoin/bitcoin/pull/16373

### CODE REVIEW

**Comment by instagibbs** [comment]
Need to change the "Send" button text to whatever the parent PR is doing

**Comment by jonasschnelli** [comment]
Concept ACK

**Comment by Sjors** [comment]
Concept ACK. Linter complains about circular dependency.

**Comment by instagibbs** [comment]
@Sjors appears to be due to the forward declaration of `SendCoinsRecipient` in `qt/guiutil.h`, and is already somewhat covered in an exception:
`"qt/guiutil -> qt/walletmodel -> qt/optionsmodel -> qt/guiutil"`

If my reading is correct I propose to add this particular path to the exception list in lieu of a larger refactor which should get rid of both the exceptions:
`"qt/guiutil -> qt/walletmodel -> qt/guiutil`

**Comment by instagibbs** [comment]
Removed the circular dependency(and removed a whitelisted one).

**Comment by instagibbs** [comment]
I'll rebase onto master once https://github.com/bitcoin/bitcoin/pull/17513#issuecomment-555654707 is merged

**Comment by instagibbs** [comment]
rebased onto master, split up into two commits for less conflict with https://github.com/bitcoin/bitcoin/pull/16373

**Comment by instagibbs** [comment]
rebased and ready for review now that RPC-version of this is merged

File: src/qt/walletmodel.cpp [hunk 1]
```diff
@@ -519,6 +522,21 @@ bool WalletModel::bumpFee(uint256 hash, uint256& new_hash)
         return false;
     }
 
+    // Short-circuit if we are returning a bumped transaction PSBT to clipboard
+    if (create_psbt) {
+        PartiallySignedTransaction psbtx(mtx);
+        bool complete = false;
+        const TransactionError err = wallet().fillPSBT(psbtx, complete, SIGHASH_ALL, false /* sign */, true /* bip32derivs */);
+        assert(!complete);
+        assert(err == TransactionError::OK);
```
Comment (Sjors) [inline src/qt/walletmodel.cpp]: Nit: check this first

**Review by Sjors — APPROVED** [review]
Tested ACK e3b19d869612b637f8bb702add0c363afe8adb8f

It would be nice if we can get the PSBT generation stuff in one place, so #17509 can build on that.

File: src/qt/walletmodel.cpp (hunk 1 above)
Comment (instagibbs) [reply src/qt/walletmodel.cpp]: for the future or in case I touch this again: why?

**Comment by achow101** [comment]
Code review ACK e3b19d869612b637f8bb702add0c363afe8adb8f

File: src/qt/walletmodel.cpp (hunk 1 above)
Comment (gwillen) [reply src/qt/walletmodel.cpp]: The value of complete isn't necessarily meaningful if err is set, I would say, so in that case it's probably less misleading to show a failure message referencing err, where one referencing complete seems potentially very misleading.

Comment (gwillen) [reply src/qt/walletmodel.cpp]: ... actually, for the record, assert()s here kind of scare me -- is that a practice we generally use? Would it be better to do as we do above, and put up a critical messagebox and return failure? Blowing up the process is drastic here; the failure seems well-contained even if surprising.

**Comment by gwillen** [comment]
Code review ACK e3b19d8, nits/nice-to-haves:
- I kind of agree with Sjors that it would be nice to end up with all the PSBT-generation stuff in one place for further development on it.
- See my comment above about the asserts.

File: src/qt/walletmodel.cpp (hunk 1 above)
Comment (instagibbs) [reply src/qt/walletmodel.cpp]: pushed a commit to address the assert concern

**Comment by instagibbs** [comment]
At the risk of invalidating 3 ACKs, I pushed a commit that converts the asserts to error message box alert instead.

**Comment by gwillen** [comment]
Thanks, code review ACK 267c77f.

File: src/qt/walletmodel.cpp (hunk 1 above)
Comment (Sjors) [reply src/qt/walletmodel.cpp]: Also because you changed that in the earlier PR too :-)

**Comment by Sjors** [comment]
Concept ACK on using `QMessageBox::critical` instead of `assert`. Are sure it's always safe to continue putting the PSBT on the clipboard after a failure?

**Comment by gwillen** [comment]
> Concept ACK on using `QMessageBox::critical` instead of `assert`. Are you sure it's always safe to continue putting the PSBT on the clipboard after a failure?

Oh, indeed, there should be a "return false;" after the messagebox. This is what you get for making reasonable changes in response to review, Greg. :-(

**Comment by instagibbs** [comment]
fixed :)

**Comment by Sjors** [comment]
utACK 3c30d71

**Comment by gwillen** [comment]
code review ACK 3c30d71

**Comment by achow101** [comment]
ACK 3c30d7118a5d5cb40c3686e0da884d3928caaeba

nit: squash, but 3 ACKs, so meh.

**Review by meshcollider — COMMENTED** [review]
We now have essentially the same code in the bumpfee RPC, in the GUI when drafting a transaction, and now here when bumping in the GUI too. It would be nice to consolidate it a bit. If you want to leave it to a follow-up though I am happy to merge this with the current ACKs.

**Review by promag — COMMENTED** [review]
Correct me if I'm wrong but I see 3 things that should be addressed __later__:
 1. `bumpFee` should not be in the wallet model - could follow same approach as `WalletControllerActivity` - because IMO it's wrong to open message boxes (any GUI change) or copy to clipboard from a model class;
 2. `bumpFee` should be asynchronous - 1. above would help here - if from the GUI thread we hit cs_main or cs_wallet then that's bad;
 3. we keep adding nested event loops with QMessageBox helper functions - again 1. would help here.

**Comment by instagibbs** [comment]
@meshcollider I'm going to claim ignorance on how to best structure QT code and ask you merge this as-is. Sorry!

**Comment by promag** [comment]
Code review ACK 3c30d7118a5d5cb40c3686e0da884d3928caaeba.

I'll address my concerns in a follow up.

**Comment by fanquake** [comment]
Given the 4 ACKs and self-confessed Qt ignorance, I'm going to merge this. Will open an issue to capture all the items that need addressing as followup.

**Pull request merged**